### ISSUE TITLE
Dirty problems using Sequel and Mobility

### ISSUE THREAD

**Issue by jnylen** [shioyama/mobility#79]
Returns an error if plugin: :dirty is loaded in Sequel even though its: dirty: true

## Context
Using Sequel with jsonb and `, dirty: true` on the translates columns.

## Expected Behavior
Well, it should work.

## Actual Behavior
In sequel.rb:
`Sequel::Model.plugin :dirty`

In model:
```ruby
  translates :title, dirty: true
  translates :synopsis, dirty: true
```

Returns (on Model.new):
```
2.3.3 :001 > a = Model.new
NoMethodError: undefined method `[]' for nil:NilClass
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/backends/hash_valued.rb:14:in `read'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/backend/stringify_locale.rb:10:in `read'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/plugins/cache/translation_cacher.rb:23:in `block in initialize'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/plugins/fallbacks.rb:99:in `block in define_read'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/plugins/presence.rb:25:in `read'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/plugins/default.rb:64:in `block in initialize'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/attributes.rb:185:in `block in define_reader'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/sequel-4.49.0/lib/sequel/plugins/dirty.rb:194:in `change_column_value'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/sequel-4.49.0/lib/sequel/model/base.rb:1468:in `[]='
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/backends/sequel/pg_hash.rb:32:in `block (4 levels) in <class:PgHash>'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/backends/sequel/pg_hash.rb:32:in `each'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/mobility-0.2.1/lib/mobility/backends/sequel/pg_hash.rb:32:in `block (3 levels) in <class:PgHash>'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/sequel-4.49.0/lib/sequel/model/base.rb:1440:in `initialize'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/enumerize-2.1.2/lib/enumerize/base.rb:51:in `initialize'
	from (irb):1:in `new'
	from (irb):1
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/hanami-1.0.0/lib/hanami/commands/console.rb:61:in `start'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/hanami-1.0.0/lib/hanami/cli.rb:89:in `console'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/gems/hanami-1.0.0/bin/hanami:5:in `<top (required)>'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/bin/hanami:22:in `load'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/bin/hanami:22:in `<main>'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:15:in `eval'
	from /home/vagrant/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:15:in `<main>'
```

Enumerize isn't the culprit, Mobility is. If I disable enumerize but keep Mobility it still fails, if I disable Dirty plugin it works, if I disable mobility but keep enumerize it works.

## Possible Fix
None

**Comment by shioyama** [comment]
Does this happen if you use Mobility version 0.1.20 ?

**Comment by jnylen** [comment]
Nope, it returns the same for 0.1.20.

**Comment by shioyama** [comment]
I can reproduce this. It seems to be a problem with the jsonb columns not having a default value. In AR this is handled by setting `default: ''` (in Rails 4 it was `default: '{}'`), which in the tests happens [here](https://github.com/shioyama/mobility/blob/ef7adc36fdb3b062a8e175aaec1c801ae9f17ed9/spec/active_record/schema.rb#L94-L100). In the case of Sequel this doesn't seem to be happening and I can't find much documentation on it.

Without a default value, this will happen when dirty tries to check the previous value of a new column.

**Comment by shioyama** [comment]
Hmm... maybe this: https://github.com/jeremyevans/sequel/issues/1330

I'll have to look into this a bit.

**Comment by shioyama** [comment]
Ok, I confirmed I can fix this with:

```ruby
self.plugin :defaults_setter
attributes.each { |attribute| self.default_values[attribute.to_sym] = lambda { ::Sequel.pg_jsonb({}) } }
```

in the `setup` block for the jsonb backend. I suppose hstore may need it as well.

**Comment by shioyama** [comment]
Try the `sequel_jsonb_dirty` branch, here: https://github.com/shioyama/mobility/tree/sequel_jsonb_dirty

**Comment by jnylen** [comment]
@shioyama Seems to work.

Thanks again for this gem! :)

**Comment by shioyama** [comment]
Great, I'll just have a closer look and merge that change in the next couple days. For now pls just use that branch.

**Comment by shioyama** [comment]
The fix for this issue has been released in version [0.2.2](https://rubygems.org/gems/mobility/versions/0.2.2).
